subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c

```diff
@@ -0,0 +1,453 @@
+					(align)) + VRING_USED_SIZEOF((vring_size)))
+/* Macro for calculating used memory by virtqueue buffers for remote device. */
+#define VIRTQUEUE_SIZE_GET(vring_size)  (RPMSG_BUFFER_SIZE * (vring_size))
+
+/* Macro for getting the size of shared memory occupied by single IPC instance. */
+#define SHMEM_INST_SIZE_GET(vring_size) (VDEV_STATUS_SIZE +  \
+					(2 * VIRTQUEUE_SIZE_GET((vring_size))) + \
+					(2 * VRING_SIZE_COMPUTE((vring_size), (VRING_ALIGNMENT))))
+
+/* Returns size of used shared memory consumed by all IPC instances*/
+#define SHMEM_CONSUMED_SIZE_GET(vring_size) (IPC_INSTANCE_COUNT * SHMEM_INST_SIZE_GET((vring_size)))
+
+/* Returns maximum allowable size of vring buffers to fit memory requirements. */
+#define VRING_SIZE_GET(shmem_size)                                       \
+					((SHMEM_CONSUMED_SIZE_GET(32)) < (shmem_size) ? 32 : \
+					(SHMEM_CONSUMED_SIZE_GET(16)) < (shmem_size) ? 16 :  \
+					(SHMEM_CONSUMED_SIZE_GET(8))  < (shmem_size) ? 8  :  \
+					(SHMEM_CONSUMED_SIZE_GET(4))  < (shmem_size) ? 4  :  \
+					(SHMEM_CONSUMED_SIZE_GET(2))  < (shmem_size) ? 2 : 1)
+
+/* Returns size of used shared memory of single instance in case of using
+ * maximum allowable vring buffer size.
+ */
+#define SHMEM_INST_SIZE_AUTOALLOC_GET(shmem_size) \
+					  (SHMEM_INST_SIZE_GET(VRING_SIZE_GET((shmem_size))))
+
+/* Returns start address of ipc instance in shared memory. It assumes that
+ * maximum allowable vring buffer size is used.
+ */
+#define SHMEM_INST_ADDR_AUTOALLOC_GET(shmem_addr, shmem_size, id) ((shmem_addr) + \
+					  ((id) * (SHMEM_INST_SIZE_AUTOALLOC_GET(shmem_size))))
+
+#if IS_ENABLED(CONFIG_RPMSG_MULTI_INSTANCE_MASTER)
+#define VIRTQUEUE_ID 0
+#else
+#define VIRTQUEUE_ID 1
+#endif
+
+
+static bool config_correct;
+static int instance;
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *p_ep)
+{
+	rpmsg_destroy_ept(p_ep);
+}
+
+static unsigned char virtio_get_status(struct virtio_device *p_vdev)
+{
+	struct  rpmsg_mi_ctx *ctx = metal_container_of(p_vdev, struct rpmsg_mi_ctx, vdev);
+
+	return (IS_ENABLED(CONFIG_RPMSG_MULTI_INSTANCE_MASTER) ?
+			VIRTIO_CONFIG_STATUS_DRIVER_OK :
+			sys_read8(ctx->shm_status_reg_addr));
+}
+
+static uint32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return BIT(VIRTIO_RPMSG_F_NS);
+}
+
+#ifdef CONFIG_RPMSG_MULTI_INSTANCE_MASTER
+
+static void virtio_set_status(struct virtio_device *p_vdev, unsigned char status)
+{
+	struct rpmsg_mi_ctx *ctx  = metal_container_of(p_vdev, struct rpmsg_mi_ctx, vdev);
+
+	sys_write8(status, ctx->shm_status_reg_addr);
+}
+
+
+static void virtio_set_features(struct virtio_device *vdev, uint32_t features)
+{
+	/* No need for implementation */
+}
+#endif
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	int status;
+
+	struct rpmsg_mi_ctx *ctx = metal_container_of(vq->vq_dev, struct rpmsg_mi_ctx, vdev);
+
+	if (ctx) {
+		status = ipm_send(ctx->ipm_tx_handle, 0, 0, NULL, 0);
```

Comment:
I strongly suggest to use a user configurable ID here. On some platforms different IDs can have a different meaning on the remote side, so better make this configurable.

Reply:
Do you mean a SOC identifier such as CONFIG_SOC_NRF5340_CPUAPP or CONFIG_SOC_NRF5340_CPUNET?

Reply:
No, I mean the ID parameter of the `ipm_send` call. You hardcoded that to `0` but that could not always be available or the correct one to use. In some cases not even all the IDs are available for IPC, so better make that configurable.